packages/server/src/api/controllers/automation.ts

```diff
@@ -231,24 +231,25 @@ export async function test(
   ctx: UserCtx<TestAutomationRequest, TestAutomationResponse>
 ) {
   const db = context.getAppDB()
-  let automation = await db.get<Automation>(ctx.params.id)
-  await setTestFlag(automation._id!)
-  const testInput = prepareTestInput(ctx.request.body)
-  const response = await triggers.externalTrigger(
-    automation,
-    {
-      ...testInput,
-      appId: ctx.appId,
-      user: sdk.users.getUserContextBindings(ctx.user),
-    },
-    { getResponses: true }
-  )
-  // save a test history run
-  await updateTestHistory(ctx.appId, automation, {
-    ...ctx.request.body,
-    occurredAt: new Date().getTime(),
+  const automation = await db.tryGet<Automation>(ctx.params.id)
+  if (!automation) {
+    ctx.throw(404, `Automation ${ctx.params.id} not found`)
+  }
+
+  const { request, appId } = ctx
+  const { body } = request
+
+  ctx.body = await withTestFlag(automation._id!, async () => {
```

Comment:
Rolled up `setTestFlag` and `clearTestFlag` into this `withTestFlag` helper that correctly handles what happens when an exception is thrown. While I was here, I did a little refactoring to make this handler easier to read.